packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+
+	// 1. generate asset manifest
+	const manifest = await walk(directory, {});
+
+	// 2. fetch buckets w/ hashes
+	const initializeAssetsResponse = await fetchResult<InitializeAssetsResponse>(
+		`/accounts/${accountId}/workers/scripts/${scriptName}/assets-upload-session`,
+		{
+			method: "POST",
+			body: JSON.stringify({ manifest: manifest }),
+		}
+	);
+
+	// if nothing to upload, return
+	if (initializeAssetsResponse.buckets.flat().length === 0) {
+		return { manifest, jwt: initializeAssetsResponse.jwt };
+	}
+
+	// 3. fill buckets and upload assets
+	const includedHashes = initializeAssetsResponse.buckets.flat();
+	const filteredFiles = Object.entries(manifest).filter((entry) =>
+		includedHashes.includes(entry[1].hash)
+	);
+	const queue = new PQueue({ concurrency: BULK_UPLOAD_CONCURRENCY });
+	let attempts = 0;
+	const start = Date.now();
+	let assetLogCount = 0;
+	let bucketUploadCount = 0;
+	let completionJwt = "";
+
+	for (const bucket of initializeAssetsResponse.buckets) {
+		attempts = 0;
+		let gatewayErrors = 0;
+		const doUpload = async (): Promise<UploadResponse> => {
+			// Populate the payload only when actually uploading (this is limited to 3 concurrent uploads at 50 MiB per bucket meaning we'd only load in a max of ~150 MiB)
+			// This is so we don't run out of memory trying to upload the files.
+			const payload: UploadPayloadFile[] = await Promise.all(
+				bucket.map(async (fileHash) => {
+					const manifestEntryIndex = filteredFiles.findIndex(
+						(file) => file[1].hash === fileHash
+					);
+					// not sure if this is really a user error - what should i use instead if so?
+					if (manifestEntryIndex === -1) {
+						throw new FatalError(
+							`A file was requested that does not appear to exist`,
+							1
+						);
+					}
+					const manifestEntry = filteredFiles.splice(manifestEntryIndex, 1)[0];
+					const absFilePath = path.join(assetDirectory, manifestEntry[0]);
+					// just logging the upload at the moment...
+					// unsure how to log deletion vs unchanged file ignored/if we want to log this
+					assetLogCount = logAssetUpload(
+						`+ ${manifestEntry[0]}`,
+						assetLogCount
+					);
+					return {
+						name: manifestEntry[0],
+						hash: fileHash,
+						contents: (await readFile(absFilePath)).toString("base64"),
+						metadata: {
+							contentType: getType(absFilePath) || "application/octet-stream",
+						},
+					};
+				})
+			);
+
+			try {
+				const res = await fetchResult<UploadResponse>(
+					`/accounts/${accountId}/workers/assets/upload`,
+					{
+						method: "POST",
+						headers: {
+							"Content-Type": "application/x-ndjson",
+							Authorization: `Bearer ${initializeAssetsResponse.jwt}`,
+						},
+						body: payload.map((x) => JSON.stringify(x)).join("\n"),
+					}
+				);
+				logger.debug("result:", res);
+				return res;
+			} catch (e) {
+				if (attempts < MAX_UPLOAD_ATTEMPTS) {
+					logger.debug("failed:", e, "retrying...");
```

Comment:
similar to https://github.com/cloudflare/workers-sdk/pull/6374/files#r1701809610